c10/core/WrapDimMinimal.h

```diff
@@ -4,37 +4,22 @@
 
 namespace c10 {
 
+namespace detail {
+C10_API int64_t
+maybe_wrap_dim_slow(int64_t dim, int64_t dim_post_expr, bool wrap_scalar);
+}
+
 static inline int64_t maybe_wrap_dim(
     int64_t dim,
     int64_t dim_post_expr,
     bool wrap_scalar = true) {
-  if (dim_post_expr <= 0) {
-    if (!wrap_scalar) {
-      TORCH_CHECK_INDEX(
-          false,
-          "dimension specified as ",
-          dim,
-          " but tensor has no dimensions");
-    }
-    dim_post_expr = 1; // this will make range [-1, 0]
-  }
-
-  int64_t min = -dim_post_expr;
-  int64_t max = dim_post_expr - 1;
-  if (dim < min || dim > max) {
-    TORCH_CHECK_INDEX(
-        false,
-        "Dimension out of range (expected to be in range of [",
-        min,
-        ", ",
-        max,
-        "], but got ",
-        dim,
-        ")");
+  // Inline the fast paths
+  if (C10_LIKELY(-dim_post_expr <= dim && dim < dim_post_expr)) {
+    // Branch-less version of dim + (dim < 0 ? dim_post_expr : 0)
```

Comment:
the two versions of clang I tried generate the same assembly (using a branch) either way -- https://godbolt.org/z/xPG8M1nWr

Reply:
The branching in that assembly is for the `if` statement above (which is likely to be predicted correctly, so okay to branch). 

Clang is making a similar optimization as me to make the ternary version branch-less
```
        mov     rax, rdi
        sar     rax, 63 
        and     rax, rsi
        add     rax, rdi
```

This extends the sign bit of `dim` to fill the entirety of `rax`, then does `dim + (dim_post_expr & rax)`.